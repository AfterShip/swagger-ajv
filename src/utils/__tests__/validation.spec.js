'use strict';

const validation = require('../validation');

const toJson = error => (
	Object
		.keys(error)
		.reduce(
			(acc, k) => ({
				...acc,
				[k]: error[k],
			}),
			{},
		)
);

describe('validation', () => {
	const components = {
		'schemas': {
			'Data': {
				'type': 'object',
			},
		},
	};

	const paths = {
		'/get': {
			'get': {
				'parameters': [{
					'in': 'query',
					'name': 'query',
					'required': true,
					'schema': {
						'type': 'string',
						'enum': ['query'],
						'errorMessage': 'This is a custom error message',
					},
				}],
			},
		},
		'/post': {
			'post': {
				'requestBody': {
					'content': {
						'application/json': {
							'schema': {
								'type': 'object',
								'properties': {
									'body': {
										'type': 'number',
										'mustPositive': true,
									},
								},
								'required': [
									'body',
								],
							},
						},
					},
				},
			},
		},
	};

	const validate = validation({
		components,
		paths,
		ajvKeywords: [{
			name: 'mustPositive',
			def: {
				validate: (schema, data) => {
					return schema ? data > 0 : true;
				},
			},
		}],
	});

	test('valid get request', () => {
		expect(
			validate({
				body: {},
				headers: {},
				method: 'get',
				params: {},
				query: {
					query: 'query',
				},
				route: '/get',
			}),
		).toEqual(undefined);
	});

	test('invalid get request', () => {
		try {
			validate({
				body: {},
				headers: {},
				method: 'get',
				params: {},
				query: {
					query: 'body',
				},
				route: '/get',
			});
			expect(true).toEqual(false);
		} catch (error) {
			expect(error.message).toMatchSnapshot();
			expect(toJson(error)).toMatchSnapshot();
		}
	});

	test('valid post request', () => {
		expect(
			validate({
				body: {
					body: 1,
				},
				headers: {},
				method: 'post',
				params: {},
				query: {},
				route: '/post',
			}),
		).toEqual(undefined);
	});

	test('invalid post request', () => {
		try {
			validate({
				body: {
					body: {
						body: 'body',
					},
				},
				headers: {},
				method: 'post',
				params: {},
				query: {},
				route: '/post',
			});
			expect(true).toEqual(false);
		} catch (error) {
			expect(error.message).toMatchSnapshot();
			expect(toJson(error)).toMatchSnapshot();
		}
	});

	test('invalid post request body <= 0', () => {
		try {
			validate({
				body: {
					body: -2,
				},
				headers: {},
				method: 'post',
				params: {},
				query: {},
				route: '/post',
			});
			expect(true).toEqual(false);
		} catch (error) {
			expect(error.message).toMatchSnapshot();
			expect(toJson(error)).toMatchSnapshot();
		}
	});
});
