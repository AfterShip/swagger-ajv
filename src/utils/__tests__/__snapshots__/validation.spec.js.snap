// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation invalid get request 1`] = `"Schema validation error"`;

exports[`validation invalid get request 2`] = `
Object {
  "ajv": Array [
    Object {
      "instancePath": "/query/query",
      "keyword": "errorMessage",
      "message": "This is a custom error message",
      "params": Object {
        "errors": Array [
          Object {
            "emUsed": true,
            "instancePath": "/query/query",
            "keyword": "enum",
            "message": "must be equal to one of the allowed values",
            "params": Object {
              "allowedValues": Array [
                "query",
              ],
            },
            "schemaPath": "#/properties/query/properties/query/enum",
          },
        ],
      },
      "schemaPath": "#/properties/query/properties/query/errorMessage",
    },
  ],
  "details": Array [
    Object {
      "info": "data.query This is a custom error message",
      "parentSchemaId": undefined,
      "path": "data.query",
    },
  ],
}
`;

exports[`validation invalid post request 1`] = `"Schema validation error"`;

exports[`validation invalid post request 2`] = `
Object {
  "ajv": Array [
    Object {
      "instancePath": "/body/body",
      "keyword": "type",
      "message": "must be number",
      "params": Object {
        "type": "number",
      },
      "schemaPath": "#/properties/body/properties/body/type",
    },
    Object {
      "instancePath": "/body/body",
      "keyword": "mustPositive",
      "message": "must pass \\"mustPositive\\" keyword validation",
      "params": Object {},
      "schemaPath": "#/properties/body/properties/body/mustPositive",
    },
  ],
  "details": Array [
    Object {
      "info": "data.body must be number",
      "parentSchemaId": undefined,
      "path": "data.body",
    },
    Object {
      "info": "data.body must pass \\"mustPositive\\" keyword validation",
      "parentSchemaId": undefined,
      "path": "data.body",
    },
  ],
}
`;
